### DISCUSSION TITLE
Issues with saved datacubes

### DISCUSSION THREAD

**Discussion by MambaYasmin** [openhsi/openhsi#29] [Q&A]
Hello,

I'm working with the LUCID Camera and have the python file (finally) working. I get the loading bar as it is capturing datacubes but I don't know where anything is saving. Please provide detailed instructions. I am also running into Save C errors as if its having a memory issue. This memory issue also appears when I use the arena view software. 
![image](https://user-images.githubusercontent.com/99000544/158656479-b82846a3-90e0-472f-b88e-0c6c8fdb753b.png)

Thanks

**Comment by christhechris** [comment]
Hi @MambaYasmin,

For the ArenaSDK install you should not need to change any thing in the Arena library to make it work. Can you let me know what wasn't working so we can update guides?

In your current code, nothing is saved, just sorted in the cam object (the raw lumpy array can be accessed at cam.dc.data).

The save method is described here: https://openhsi.github.io/openhsi/data.html#DataCube.save You use it like: 
`cam.save(save_dir='data_dir', prefix='some_prefix', suffix='some_suffix')`


PS- If you can start questions on the Robonation forum, that would be awesome (they just want to make sure any issues or visible to other teams in case they have similar issues etc).

**Comment by MambaYasmin** [comment]
Thanks for the reply (meant to jump back on this quicker)

I have started the save code: To be clear Photos taken on 03/24/2022
save_dir="hsigang"
some_prefix= "Mamba_Y"
some_suffix="Photos"

I'd have on my computer saved **C:\hsigang\Mamba_Y.Photos?**

Thanks!
![image](https://user-images.githubusercontent.com/99000544/160000277-42657ea6-9f26-4c43-bfd7-19ded1966ffa.png)

**Reply by christhechris** [reply]
you appear to be hitting some logic included to check available memory on your device (by default if the data cube you are trying to collect will use more then 80% free memory it asked you if you want to continue).

You can disable this by using the the `warn_mem_use=False` or asking for fewer lines, say 500, instead.

See https://openhsi.github.io/openhsi/data.html#DataCube

FYI - there have been a few other bug fixes in a recent updates so you may want to update your install as well (`conda update openhsi` or `pip install openhsi --upgrade`).

FYI #34 @YiweiMao

**Reply by christhechris** [reply]
Also did the code ask for an input? It should have asked you to enter `y/n`, and anything other then `y` would end the run.

it looks like maybe they way the code was entered that `cam.show(...)` was intercepted as the input maybe? You might want to try ipython prompt or Jupiter notebook instead (at least for early testing).

**Reply by MambaYasmin** [reply]
(I will post this chain on the forum once I get through my mysteries)
Okay I updated my open hsi and now I'm having syntax errors from the data.py on the line pictured. I have not editied the source data. py file at all. This is done with the simulated camera but I also have the identical issue with the lucid camera version when I have the camera plugged in.

![image](https://user-images.githubusercontent.com/99000544/161451309-3479024b-2aa0-4c37-b679-35ae5df10080.png)

Additionally I have gotten errors saying 'int has no object has no attribute strftime' when I run the code with the lucid lense set up.
In the black highlight is the syntax error highlighted line 462. And the no attribute error is in response to lines 4441 and 442

![image](https://user-images.githubusercontent.com/99000544/161451407-0eb9b458-b504-4466-bad6-85bdee0e1ff4.png)
 I appreciate the help so far.

**Reply by christhechris** [reply]
Hi @MambaYasmin I'm thinking this is something to do with python version and dependancies.
A couple of questions to help us diagnose things:

1. I see you are running on windows currently, how have you installed python and which version? 
2. Are you running from a version of OpenHSI installed from pip or conda or using the git repo?
3. can also you show the output of 
`python -VV`
and
`pip list` or `conda list` depending which you use.


Another thing to try is to start from scratch with the new install guide i.e. https://openhsi.github.io/openhsi/tutorial_installing_windows.html

**Reply by christhechris** [reply]
@YiweiMao any other thoughts?

**Reply by MambaYasmin** [reply]
1. I installed python 3.8
2. Openhsi is with the upgrade command you mentioned. I did it in my command line.
3. 
![image](https://user-images.githubusercontent.com/99000544/161453071-fe63a933-39b6-4818-958e-e0a0502beeb1.png)
![image](https://user-images.githubusercontent.com/99000544/161453084-00f8f682-7f6e-4dca-96c7-f619fcd854cf.png)
![image](https://user-images.githubusercontent.com/99000544/161453092-b2152112-714d-4cf8-bd53-bea929d69e3c.png)

**Reply by christhechris** [reply]
Lets try to uninstall and reinstall of open hsi.

`pip uninstall  openhsi`
`pip install openhsi`


Can you also show output of to show which distribution of python you have
`python -VV`

**Reply by YiweiMao** [reply]
Hmm, the SyntaxError during import doesn't tell much, it could be the preceding lines but I've tested this on Python 3.8 and it runs fine for me - could be some install issue. 
As for the `strftime` issue, can you run this snippet and tell us the output? (This is a minimal example of what `DateTimeBuffer` implements)
```python
from datetime import datetime, timezone, timedelta
ts = datetime.timestamp(datetime.now())
dt = datetime.fromtimestamp(ts, tz=timezone.utc)
print(f"Current UTC date is {dt.strftime('%Y_%m_%d')}")
```

Note: A recent pull request fixed an issue *saving* datacubes collected with slow binning (`processing_lvl=3`) https://github.com/openhsi/openhsi/pull/35 This was not a SyntaxError though. 
We'll release this version on PyPi/Conda soon.

**Reply by christhechris** [reply]
I've just replicated your code in a clean Miniforge install on windows (following the install guide) and I noticed your import of the SimulatedCamera is
`from openhsi.cameras import SimulatedCamera`

when I think it should be

`from openhsi.capture import SimulatedCamera`

With that import change everything works as expected. You version should be giving an import error, not a syntax error, in any case can you try making this change as well.

In the screen cap it also looks like there is an extra space in front of the import, but that should given an indent error not the syntax one you show, but lets confirm that is not the source as well.

edit* couple clarification to the above.

**Comment by MambaYasmin** [comment]
![image](https://user-images.githubusercontent.com/99000544/161775737-cfb7de78-c489-4ba4-b282-35ff90e30dd5.png)

If I run the snippet in command line and then in my python terminal.

![image](https://user-images.githubusercontent.com/99000544/161775768-33a2a37a-42c3-4499-a514-2b2ab92bad5f.png)


Both outputs are as you mentioned. I will give it try with a fresh install.  However I have re-installed the arena view api, the sdk, python and open hsi once if not twice all at this point.

Let me know please

Yasmin
________________________________

**Reply by christhechris** [reply]
It seems like images did not come through, can you repost them?

**Reply by MambaYasmin** [reply]
Added the pics
UPDATE: I got the simulated camera running in jupyter now. Ultimately this has to run on a Jestson nano. Should I expect to debug in jupyter and then be able the run the the python files through command line as a typical script? Once it work on Jupyter with my Lucid Lens?

Thanks

**Reply by christhechris** [reply]
If the Jupyter notebook versions is working locally (either simulated camera or the Robonation OpenHSI) I would expect the same or similar code to work via a script as well. The only exceptions are some of the plotting tools assume they are in notebook environment, but these are not required to acquire and process data etc.

One thought we had that may be cause the issue you saw, is if you have cloned the entire OpenHSI repo (i.e. too get the assets folder) and are running code located in that folder, then the local version of the OpenHSI library (which may have been unintentionally edited) may be taking precedence over an installed version. This would be constant with reinstalling/upgrading etc not fixing things.

**Comment by YiweiMao** [comment]
Here is a minimal example I created on Google Colab demonstrating `SimulatedCamera`.
https://colab.research.google.com/drive/1VEzXZO0DCbh-L5vNJ9c5fOCo7osSL4pK?usp=sharing

You can test it in your own Colab session is you like. 

We also have an example on the homepage demonstrating `SimulatedCamera`
https://openhsi.github.io/openhsi/#How-to-use
You should use your own image, settings, and calibration files.

**Reply by MambaYasmin** [reply]
So I have gone on to another step. Thanks for the help previously. I am moving on to process my data cubes into reflectance but am getting the error when running in jupyter. Please let me know what you think. This error only appears at processing_lvl high than 2

![image](https://user-images.githubusercontent.com/99000544/177017345-02494126-b4f6-4470-be5f-eb063ff27b65.png)

**Reply by MambaYasmin** [reply]
And this is is the code I'm working with.
Additionally I understand the processing levels are to give us different values but does the "reflectance" output an array over the wavelength or create an excel file with all those values in it (over the wavelengths the camera can see) ? Or am I not understanding reflectance.
![image](https://user-images.githubusercontent.com/99000544/177017476-561f7e0c-3b2a-4a27-b12d-61b460ea9f3b.png)

**Reply by christhechris** [reply]
Hi Yasmin, I think some of the error is being cut off, so it is hard to see what the root cause is. Could you repost?

**Reply by YiweiMao** [reply]
If you want radiance or reflectance (processing level >= 4) then it requires things inside the calibration file. Could be your `xarray` install version but can't tell for sure. Can you include the entire error message?
Note: For reflectance, you'll need to first precalculate the solar radiance using Py6S (https://openhsi.github.io/openhsi/tutorial_camera.html#Getting-surface-reflectance) based on your geolocation, time, etc.

**Reply by MambaYasmin** [reply]
Here is the entire error. At the very bottom I think I see the xarray install comment. 
![image](https://user-images.githubusercontent.com/99000544/177059631-555db0b9-d3fa-4c98-b5aa-81ad88c819b8.png)
This second image is a previous run of the code with processing level 4
![image](https://user-images.githubusercontent.com/99000544/177059670-4b46f030-d8f0-4c7e-823f-aa4ec6c38ae0.png)
![image](https://user-images.githubusercontent.com/99000544/177059658-c7698bf0-3334-457f-8793-b389cefb7daa.png)

**Reply by christhechris** [reply]
Hi Yasmin,

so on issue 1, to convert to reflectance you need to provide information about the illumination of the target (i.e. how much sunlight is hitting the thing being observed). This is an important concept in remote sensing/hyper spectral imaging, because you can compare reflectances taken at different times/conditions, but not radiances. This information gets stored in the rad_fit key of the calibration file, but is blank by default.

The link Yiwei shared [(here)](https://openhsi.github.io/openhsi/tutorial_camera.html#Getting-surface-reflectance)) the 6SV example is the starting point for this. There are other ways to do this empirically with known reflectance calibration targets.


on issue 2, to process previous captured data you need to specify the `.nc` file saved. Also make sure you only run processing steps not run. for example if you saved raw data, you don't want to use `.load_next_tfms` method. You only use that if you captured at say level 2 and only need to convert to radiances.

It might make sense to continue these to threads in discussion threads, to make things a bit more discoverable for others etc.

**Reply by YiweiMao** [reply]
There's two pieces of code here. One is to setting the real-time processing level to reflectance and the other is to post-process an already collected datacube. 

The first error is because it's trying to find the precalculated expected solar radiance inside the calibration file and cannot find it. To estimate the surface reflectance (0-100%), the code divides the *measured radiance* by the *expected at sensor radiance*. As you can imagine, the expected at sensor radiance depends on the Sun and sensor geometry, as well as the atmosphere. So this depends on your location and time of day! You need to precompute this if you want to use this in real-time. [This code snippet](https://openhsi.github.io/openhsi/tutorial_camera.html#Updating-the-radiative-transfer-model) shows you how to do this. You can update this without the camera attached to your computer if you wish - just load and dump using the Python standard `pickle` library (protocol=4 recommended). 
The other option is to do this empirically and estimate the reflectance based on known spectral targets. This software library can only do this on an *existing radiance datacube* (interactive and in post-processing) provided a spectral library (the spectra of your known spectral targets). 

Your 2nd error is on further *post-processing* an existing datacube. You need to provide the path to the NetCDF file (not a directory). This has a file extension of `.nc`. 
https://openhsi.github.io/openhsi/data.html#DataCube.load_nc
Finally, a note about your code snippet copied from
https://openhsi.github.io/openhsi/capture.html#ProcessDatacube
The `processing_lvl` initialises the data structures needed for that amount of processing (https://openhsi.github.io/openhsi/tutorial_camera.html#Processing-levels). So the value you set `4` is saying, initialise the functions to: crop, fast smile correction, fast binning, and conversion to radiance. 
Then the processing pipeline is modified with `ProcessDatacube.load_next_tfms` so the only thing to do, according to the code snippet copied, is to convert from digital numbers to radiance. In this example, you can see the input datacube has already been processed to level 2. 

Hope this helps you understand what this library expects! We thought long and hard coming up with these design decisions.

**Comment by MambaYasmin** [comment]
Thank you so much for the detail, definitely helpful

The only thing I could not access was the first link about real-time calculations. I get a error (may just be my university wifi) if you could even just screen shot me the code and I have to re-type it that's okay.

Thanks for everything
Yasmin

[cid:ca8db794-116d-4234-8dbe-601b20c384a0]
________________________________

**Reply by YiweiMao** [reply]
The Radiative Transfer stuff for predicting reflectance is here in the Quick Start Guide
https://openhsi.github.io/openhsi/tutorial_camera.html#Getting-surface-reflectance

I was thinking of updating the guide here so you can add to the calibration file without having a camera connected and initialised. 
Here's a code snippet to do this using the base class `CameraProperties` (that your camera class inherits)

```python
#other imports. e.g. interp1
from openhsi.atmos import *
from openhsi.data import CameraProperties
cam = CameraProperties(json_path="your_settings_file.json",  pkl_path="your_calibration_file.pkl")

# run the Py6S model using `Model6SV`
# cam.calibration["rad_fit"] = model radiance / 10
# cam.dump(json_path,pkl_path) # update the settings and calibration files
```
